### PR TITLE
SEP-31: integrate SEP-38

### LINKED ISSUE #916: SEP-31: Integrate SEP-38 RFQ Quotes
Now that we have a standard for providing clients quotes between on and off chain assets, we should integrate this functionality into SEP-31.

The basic idea is that the sending anchor would send an on-chain asset A and the receiving anchor would deliver an off-chain asset B at a rate the receiving anchor provided the sending anchor in a previous request.

### PR DESCRIPTION
**Pull request by JakeUrban**
Resolves #916

Integrates SEP-38 into SEP-31.

### Changes

- Clarifies the assets listed in a `GET /info` response to be the Stellar assets the Receiving Anchor accepts from a Sending Anchor. This makes the API consistent with SEP-24 and SEP-6
- Adds a "Asset Conversions and Liquidity Strategies" section outlining the three methods possible for sending and converting assets
- Adds the following parameters to `POST /transactions`:
    - `destination_asset`: used when the Receiving Anchor does not support providing firm quotes
    - `quote_id`: used when the Receiving Anchor supports firm quotes
- Makes the following `POST /transactions` parameters optional:
    - `amount_in`
    - `asset_code`
        - neither of these parameters are needed when `quote_id` is provided
- Adds the following attributes to the `GET /transactions/:id` response:
    - `amount_in_asset`
    - `amount_out_asset`
    - `amount_fee_asset`

### CODE REVIEW

File: ecosystem/sep-0031.md
Comment (tomerweller) [inline ecosystem/sep-0031.md:94]: This is really great. We should have @phil-stellar and @lisaclaire828 take a look as well. I would consider the following renaming of strategies to reflect *when* the asset conversion happens: 

`Sending Anchor Asset Conversion` -> `Pre-payment asset conversion`

`Stellar Network Asset Conversion`-> `Real time asset conversion` 

`Receiving Anchor Asset Conversion` -> `Post-payment asset conversion`

File: ecosystem/sep-0031.md
Comment (leighmcculloch) [inline ecosystem/sep-0031.md:73]: 💡 This paragraph is a little difficult to digest because it leads with the least important information and finishes with the most important information. The most important information is what the protocol is intended to be used for. Suggest reordering:
```suggestion
This protocol is designed for the common case where the Sending Client provides the Sending Anchor with a particular asset (the source asset) and the Receiving Client receives a different asset (the destination asset). It is possible to use this protocol when the Sending and Receiving clients wish to send and receive the same asset, but that is not the common use case it has been designed.
```

File: ecosystem/sep-0031.md
Comment (leighmcculloch) [inline ecosystem/sep-0031.md:75]: >In this case,

💡  It's ambiguous at the start of the sentence what "this case" is because the prior paragraph discusses two cases. The reader has to look ahead towards to rest of the paragraph to figure it out. I think this would make the paragraph clearer:

```suggestion
When the source asset and destination assets are different, the source asset must convert into the receiving asset at some point prior to delivering funds to the Receiving Client. This conversion must be done by either the Sending Anchor, the Stellar Network, or the Receiving Anchor. The subsections below outline the strategies used for each method.
```

🔠 I also fixed a typo in this paragraph.

File: ecosystem/sep-0031.md
Comment (leighmcculloch) [inline ecosystem/sep-0031.md:85]: 🔠  There's a double instead in this sentence, and a typo.
```suggestion
Instead of holding a balance of the destination asset on Stellar, the Sending Anchor can use Path Payments to convert the source asset to the destination asset on the Stellar Network. 
```

File: ecosystem/sep-0031.md
Comment (leighmcculloch) [inline ecosystem/sep-0031.md:87]: >Note that the exchange rate between the two assets is dynamic due to market fluctuations on the SDEX.

❓ This is mentioned here as if it is unique to the SDEX, but in all cases the rate between two assets will be fluid. What's the intention with calling it out for only the SDEX?

File: ecosystem/sep-0031.md
Comment (leighmcculloch) [inline ecosystem/sep-0031.md:93]: 💭 There are some subtleties regarding submitting transactions, the created at timestamp, and how long they are valid and can be resubmitted that we might need to consider in this scenario.

The transaction not only needs to be submitted before the quote's expiration, but it needs to be included in a closed ledger. The created at timestamp on transactions is the nomination time of the ledger (the time the ledger closed) that the transaction was included in. It will be possible for a transaction to be queued, and so it could be submitted prior to the expiry time but not included in the next ledger but rather a future ledger until after the close time. This may be confusing to implementers. This is an edge case, but we should probably still document what the two parties should do if this happens. Implementers could also reduce the chance of this happening by setting the time bounds on the transaction to be the expiry time or earlier.

File: ecosystem/sep-0031.md
Comment (leighmcculloch) [reply ecosystem/sep-0031.md:94]: I think those names somewhat obfuscate who is responsible for asset conversion, and if I read them without context of what's in each section I don't really know what they mean.

**Review by leighmcculloch — COMMENTED** [review]
I've left some comments inline.

 - Typos 🔠 
 - Suggestions 💡 I defer to you on, they help me with reading it, but may be subjective.
 - Questions ❓ I defer to you on.
 - Thoughts 💭 about an area of transaction submission that I don't think this addresses. I think a change is needed, but I defer to you if this is needed right now.

File: ecosystem/sep-0031.md
Comment (JakeUrban) [reply ecosystem/sep-0031.md:93]: > Implementers could also reduce the chance of this happening by setting the time bounds on the transaction to be the expiry time or earlier.

If they do this, there is actually _no_ chance a transaction would be submitted earlier and included later than the expiration right? That detail may be important enough to explicitly mention here.

Comment (leighmcculloch) [reply ecosystem/sep-0031.md:93]: I think it's a little fuzzy, I'd verify it with the core team, but essentially yes.

File: ecosystem/sep-0031.md
Comment (JakeUrban) [reply ecosystem/sep-0031.md:87]: Fair point. I can remove it 👍

File: ecosystem/sep-0031.md
Comment (JakeUrban) [reply ecosystem/sep-0031.md:94]: I agree with Leigh, I think the _who_ is more important than the _when_

File: ecosystem/sep-0031.md
Comment (marcelosalloum) [inline ecosystem/sep-0031.md:220]: > If the Receiving Anchor does not provide exchange rates for an asset, it is assumed that the Stellar asset can be exchanged 1-for-1 with the corresponding off-chain asset.

Is that safe to assume? Are there anchors in the ecosystem operating that way?

Comment (JakeUrban) [reply ecosystem/sep-0031.md:220]: All existing anchors operate this way, since the Stellar asset they use is pegged to the fiat currency their users have.

**Review by marcelosalloum — COMMENTED** [review]
This plan sounds good but there are a few details missing IMO:

In SEP-38 how will we indicate the anchor supports the SEP? I don't see any proposed toml parameter [there](https://github.com/marcelosalloum/stellar-protocol/blob/97dd8972672d0c44eea5f6ee3a1c3d3663b3b25f/ecosystem/sep-0038.md).

If a Receiving Anchor's toml file indicates both SEP-38 and SEP-31 are implemented separately, how do we guarantee SEP-31 is supporting SEP-38? According to this proposal the Sending Anchor would need to submit a `POST /transactions` request with `quote_id` to check if it succeeds.

**Comment by JakeUrban** [comment]
@marcelosalloum responding to your comments:

> In SEP-38 how will we indicate the anchor supports the SEP? I don't see any proposed toml parameter there.

Good point. We can add a prerequisites section like we do in SEP-31 mentioning the need to have the `ANCHOR_QUOTE_SERVER` attribute in the stellar.toml, which is already defined in SEP-1.

> how do we guarantee SEP-31 is supporting SEP-38?

Thats a good question, I think there are actually two questions here:
- Does SEP-31 support SEP-38 quotes at all?
- For what assets does SEP-31 support using quotes for?

Personally, I don't see why an anchor would support SEP-38 but not support quotes to be used in the context of SEP-31 transactions. This was an assumption I made when writing this PR: if the anchor has an `ANCHOR_QUOTE_SERVER` attribute in the toml (indicating SEP-38 support), then the quotes returned from that protocol could be used in any other protocol also defined (i.e. SEP-31 and SEP-6). Maybe this is not a safe assumption to make, in which case we could add some kind of `quotes_supported: true` attribute to the SEP-31 and SEP-6 `/info` endpoints.

But then there is also the question of whether a particular asset pair is supported for exchange. For example, if I list USDC<-->BRL in my SEP-38 `GET /prices` endpoint, and I list USDC in my SEP-31 `/info` endpoint, is it possible that the anchor doesn't accept quote `id`'s for `USDC-->BRL` in a SEP-31 transaction? I don't see why that would be the case. If we wanted to be careful, we would either need to list the assets that could be exchanged for USDC in the `/info` endpoint, effectively duplicating the info in SEP-38. Or, we could adjust SEP-38 to list the protocols that can use quotes between each asset pair.

**Comment by marcelosalloum** [comment]
> Personally, I don't see why an anchor would support SEP-38 but not support quotes to be used in the context of SEP-31 transactions.

My concern is if SEP-38 is available through `ANCHOR_QUOTE_SERVER`, should we assume all other protocols that can implement SEP-38 are already implementing it? I believe an anchor should be able to support SEP-31 (with quotes) and  SEP-6 (without quotes), for instance. A boolean like `quotes_supported: true` would do the trick.

**Comment by JakeUrban** [comment]
@marcelosalloum I added `quotes_supported` as an attribute under each asset listed in `/info`. I also clarified the section header and `fee` attribute as related to the stellar asset, not the fiat currency delivered to the receiving client.

**Review by marcelosalloum — APPROVED** [review]
LGTM!

**Comment by marcelosalloum** [comment]
I'm having second thoughts on the approaches based on quotes... when posting to `/transactions` should we suggest that the Receiving Anchor could optionally implement the `quotes_supported` approach but mandatorily implement the older approach? 

Or should we make it clear that the Sending anchor should follow the `quotes_supported` approach if `quotes_supported:true`, otherwise fall back to the older SEP-31 approach?

My concern here is we could have Sending Anchors trying one approach only while Receiving Anchors are implementing the other approach, thus making their communication inconsistent.

**Comment by marcelosalloum** [comment]
Cool, using `quote_required` makes more sense to me too @JakeUrban. This way it's clearer which kind of implementation should be used for each case

File: ecosystem/sep-0031.md
Comment (marcelosalloum) [inline ecosystem/sep-0031.md:351]: 🤔  Should we be including this `created_at` attribute you're referring to? Or should it actually be `started_at`?

```suggestion
`quote_id` | string | The `id` returned from a `SEP-38 POST /quote` response. If this attribute is specified, none of the attributes listed above should also be included in the request. The Stellar transaction used to send the asset _must_ have a `started_at` timestamp that is earlier than the quote's `expires_at` attribute.
```

**Review by marcelosalloum — COMMENTED** [review]
Added a couple more suggestions below.

File: ecosystem/sep-0031.md
Comment (JakeUrban) [reply ecosystem/sep-0031.md]: I'm referring to the `created_at` attribute returned from Horizon's `GET /transactions/:id` endpoint. This value is the close time of the ledger the transaction was included in. This gives us a source of trust to use so that there is never a dispute about whether the payment was made prior to the quote's expiration.

**Comment by JakeUrban** [comment]
Hello @leighmcculloch & @marcelosalloum, now that we've settled on the technical design decisions for #917, I'd like to take what we've discussed and adjusted there and apply it to this PR. It seems like the primary decision to be made relates to whether or not we should create a new endpoint for the quote-style transactions as we did for SEP-6.

This SEP also doesn't have a `/fee` endpoint so we don't need to worry about this.

File: ecosystem/sep-0031.md
Comment (marcelosalloum) [inline ecosystem/sep-0031.md:434]: I believe the idea now is all conversion fees are included in the quote returned from SEP-38 /price and /quote endpoints, right? In that case, we don't need the attribute `amount_fee_asset` here.

File: ecosystem/sep-0031.md
Comment (marcelosalloum) [inline ecosystem/sep-0031.md:90]: Q: is this type of conversion a new addition in this version of the protocol? How many Anchors (ballpark estimation) do you think are using SEP-31 already?

**Review by marcelosalloum — COMMENTED** [review]
TBH I don't feel very strongly towards creating a new /transactions-exchange endpoint for SEP-31. Looking back at SEP-6, the meaning of `asset_code` was pretty confusing for deposits and withdrawals with non-equivalent assets but that's not the case here.

Both `asset_code` and `destination_asset` are clear concepts in SEP-31 so I think it's fine to handle all three use cases with the same existing endpoint.

---

Also, I've left a few suggestions below.

File: ecosystem/sep-0031.md
Comment (JakeUrban) [reply ecosystem/sep-0031.md:433]: I think it's worth keeping as an optional attribute in cash a fixed fee is also charged on top of the margin included in the exchange rate. If only the margin on the exchange rate is used, then `amount_fee` will be 0 anyway.

File: ecosystem/sep-0031.md
Comment (JakeUrban) [reply ecosystem/sep-0031.md:90]: Yes, prior to the development of SEP-38 and these changes to SEP-31 it wasn't possible for the receiving anchor to convert assets in a standardized way. Currently, every anchor that implements SEP-31 is interested in this capability, with many more in the pipe.

File: ecosystem/sep-0031.md
Comment (JakeUrban) [reply ecosystem/sep-0031.md:131]: I don't feel strongly either way. If we made asset & amount mandatory in all cases, the other optional parameters would be `quote_id` and `destination_asset`, which would be mutually exclusive. I don't like the idea of also requiring `destination_asset` in all requests since requiring a new field is a breaking change for existing implementations.

Comment (marcelosalloum) [reply ecosystem/sep-0031.md:131]: On second thought, requiring the user to repeat the fields that are already part of the quote goes against the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle. 

@leighmcculloch, @jacekn: if you both agree with that I'll approve this PR and update #973 to use the same strategy [here](https://github.com/stellar/stellar-protocol/pull/973/#discussion_r674033709).

Comment (leighmcculloch) [reply ecosystem/sep-0031.md:131]: It may be simpler, I think, to keep the API stable regardless if a quote is used, and to have the quote ID be an optional parameter. DRY isn't a truth, just a tool, and doesn't necessarily result in the simplest to understand APIs. In this specific case I do not feel strongly though and defer to you both.

Something that some APIs do to separate these changing input parameters from other parameters is to wrap them in a subfield and introduce a concept to wrap up those changing inputs. A good example of this is Stripe's source/payment intent interface. We could do something similar by saying the changing fields are the ExchangeParams, and for taht field you can either provide RawExchangeParams or QuoteExchangeParams.

Comment (JakeUrban) [reply ecosystem/sep-0031.md:131]: I really like this idea of a subfield, especially because the SEP-31 endpoint accepts `application/json` in a POST request. This pattern doesn't work well with SEP-6's `GET /*-exchange` endpoints though.

Comment (leighmcculloch) [reply ecosystem/sep-0031.md:131]: Even if we don't use a subfield, thinking about it more the subfield identifies a conceptual oddity with using mutual exclusion on destination_asset. For example:

<table>
<tr><th>non-quote</th><th>quote</th></tr>
<tr>
<td>

```json
{
  "asset": "...",
  "amount": "...",
  "subfield": {
    "destination_asset": "...",
  },
}
```

</td>
<td>

```json
{
  "asset": "...",
  "amount": "...",
  "subfield": {
    "quote_id": "...",
  },
}
```

</td>
</tr>
</table>

It's not intuitive to me that destination asset is inside the subfield when asset is outside. In the same way it's not intuitive that destination_asset is optional when asset isn't.

Comment (JakeUrban) [reply ecosystem/sep-0031.md:131]: Since we're not creating a `POST /transactions-exchange` endpoint and instead using the original `POST /transactions` endpoint we cannot require `destination_asset` on every request, otherwise that would be a breaking change for existing applications.

Personally I don't think having two mutually exclusive parameters is too confusing, especially if we have a subsection below the request parameters list explaining when to use each one.

Comment (JakeUrban) [reply ecosystem/sep-0031.md:131]: Updated language:

> The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide either, but not both, a `quote_id` or `destination_asset` attribute if [SEP-38](sep-0038.md) is supported.

Comment (JakeUrban) [reply ecosystem/sep-0031.md:131]: Actually, I just realized that `destination_asset` can be specified if `quote_id` is also specified. There is no need for mutual exclusion.

So the three combinations are:
- `asset_code`, `amount`
- `asset_code`, `amount`, `destination_asset`
- `asset_code`, `amount`, `destination_asset`, `quote_id`

And if `quote_id` is specified, the information must match that of the quote.

Thoughts?

**Review by marcelosalloum — APPROVED** [review]
LGTM!

**Pull request merged**
